Demo/src/main/java/com/paypal/android/viewmodels/PayPalNativeViewModel.kt

```diff
@@ -0,0 +1,170 @@
+                else -> ViewState.CheckoutError(message = error.errorDescription)
+            }
+            internalState.postValue(errorState)
+        }
+
+        override fun onPayPalCheckoutCanceled() {
+            internalState.postValue(ViewState.CheckoutCancelled)
+        }
+
+        override fun onPayPalCheckoutShippingChange(
+            shippingChangeData: ShippingChangeData,
+            shippingChangeActions: ShippingChangeActions
+        ) {
+            // implement
+        }
+    }
+
+    private val internalState = MutableLiveData<ViewState>(ViewState.Initial)
+    val state: LiveData<ViewState> = internalState
+
+    lateinit var payPalClient: PayPalClient
+
+    fun billingAgreementCheckout() {
+        viewModelScope.launch {
+            val order = getBillingAgreementTokenUseCase()
+            order.id?.also { orderId ->
+                startCheckoutFlow(CreateOrder { createOrderActions ->
+                    createOrderActions.setBillingAgreementId(orderId)
+                })
+            }
+        }
+    }
+
+    fun fetchAccessToken() {
+        internalState.postValue(ViewState.GeneratingToken)
+        viewModelScope.launch {
+            try {
+                val accessToken = getAccessTokenUseCase()
+                initPayPalClient(accessToken)
+                internalState.postValue(ViewState.TokenGenerated(accessToken))
+            } catch (e: HttpException) {
+                internalState.postValue(ViewState.CheckoutError(message = e.message))
+            }
+        }
+    }
+
+    fun orderCheckout() {
+        val order = getOrderUseCase()
+        startCheckoutFlow(CreateOrder { createOrderActions ->
+            createOrderActions.create(order) {
+                internalState.postValue(ViewState.OrderCreated(it))
+            }
+        })
+    }
+
+    fun orderIdCheckout() {
+        viewModelScope.launch {
+            val orderId = getOrderIdUseCase()
+            orderId?.also {
+                startCheckoutFlow(CreateOrder { createOrderActions ->
+                    createOrderActions.set(it)
+                    internalState.postValue(ViewState.OrderCreated(it))
+                })
+            }
+        }
+    }
+
+    fun vaultCheckout() {
+        viewModelScope.launch {
+            val sessionId = getApprovalSessionIdActionUseCase()
+            sessionId?.also {
+                startCheckoutFlow(CreateOrder { createOrderActions ->
+                    createOrderActions.setVaultApprovalSessionId(it)
+                })
+            }
+        }
+    }
+
+    private fun startCheckoutFlow(createOrder: CreateOrder) {
+        payPalClient.startCheckout(createOrder)
+    }
+
+    private fun initPayPalClient(accessToken: String) {
+        payPalClient = PayPalClient(
+            getApplication(),
+            CoreConfig(accessToken),
+            payPalConstants.returnUrl
+        )
+        payPalClient.listener = payPalListener
+    }
+
+    sealed class ViewState {
```

Comment:
this looks like MVI, in that case, LGTM is this the way forward for demo app?
`NativeCheckoutViewState` to be specific on the class and can be in different file

Reply:
Honestly i took this from the nxo repo as refernence. I dont mind whatever pattern we decide. And i dont think we should decide on one specific pattern, perhaps for diferent demos we could try diferent patterns and then decide which one is better.

Reply:
What are some alternatives to this pattern?

Reply:
well technically MVVM is diferent, you post "data" to the view, not an entire `ViewState`. In MVI, the view subscribes to one stream of view states and thats it. On MVVM, you subscribe to diferent data streams

Reply:
Yeah there are so many different patterns on Android. MVP was the main recommendation not too long ago. I'd vote for whatever scales the best. As we add more features to Northstar maintenance costs will go up in the Demo app.